src/main/java/org/lantern/proxy/GetModeProxy.java

```diff
@@ -21,37 +21,81 @@
  */
 @Singleton
 public class GetModeProxy extends AbstractHttpProxyServerAdapter {
-    private static final Logger LOGGER = LoggerFactory.getLogger(GetModeProxy.class);
-    
+    private static final Logger LOGGER = LoggerFactory
+            .getLogger(GetModeProxy.class);
+
     private final Flashlight fl;
     private final Model model;
-    
+    private volatile boolean started;
+    private volatile boolean hasPublicIp;
+
     @Inject
     public GetModeProxy(
             Model model) {
         this.model = model;
         Properties props = new Properties();
-        props.setProperty(Flashlight.CLOUDCONFIG_KEY, S3Config.DEFAULT_FLASHLIGHT_CLOUDCONFIG);
-        props.setProperty(Flashlight.CLOUDCONFIG_CA_KEY, S3Config.DEFAULT_FLASHLIGHT_CLOUDCONFIG_CA);
+        props.setProperty(Flashlight.CLOUDCONFIG_KEY,
+                S3Config.DEFAULT_FLASHLIGHT_CLOUDCONFIG);
+        props.setProperty(Flashlight.CLOUDCONFIG_CA_KEY,
+                S3Config.DEFAULT_FLASHLIGHT_CLOUDCONFIG_CA);
         fl = new Flashlight(props);
         Events.register(this);
     }
-    
+
     @Override
-    synchronized public void start() {
-        fl.startStandaloneClient();
-        fl.addFallbackProxies(model.getS3Config().getFallbacks());
-        Events.asyncEventBus().post(new ProxyConnectionEvent(ConnectivityStatus.CONNECTED));
+    public void start() {
+        boolean started = false;
+        boolean hasPublicIp = false;
+        synchronized (this) {
+            started = this.started = true;
+            hasPublicIp = this.hasPublicIp;
+        }
```

Comment:
>> I didn't make the whole methods synchronized because I don't want the delivery of the event blocking on starting flashlight.

> That won't matter because the event is delivered asynchronously on its own thread.

Okay, I went ahead and made the whole methods synchronized.